src/charm.py

```diff
@@ -125,14 +127,81 @@ def copy_stored_dict(stored_dict: StoredDict) -> dict:
         return dict(stored_dict.items())
 
+    @property
+    def auth(self) -> str:
+        """Return stored Ceph auth mode from ceph-client relation"""
+        return self._stored.ceph_data.get("auth")
+
+    @property
+    def key(self) -> str:
+        """Return stored Ceph key from ceph-client relation"""
+        return self._stored.ceph_data.get("key")
+
+    @property
+    def mon_hosts(self) -> List[str]:
+        """Return stored Ceph monitor hosts from ceph-client relation"""
+        return list(self._stored.ceph_data.get("mon_hosts", []))
+
+    def install_ceph_common(self) -> None:
+        """Install ceph-common apt package"""
+        subprocess.check_call(["apt-get", "update"])
+        subprocess.check_call(["apt-get", "install", "-y", "ceph-common"])
+
+    def write_ceph_cli_config(self) -> None:
+        """Write Ceph CLI .conf file"""
+        ceph_conf_path = "/etc/ceph/ceph.conf"
+        ceph_conf = f"""[global]
+auth cluster required = {self.auth}
+auth service required = {self.auth}
+auth client required = {self.auth}
+keyring = /etc/ceph/$cluster.$name.keyring
+mon host = {" ".join(self.mon_hosts)}
+
+log to syslog = true
+err to syslog = true
+clog to syslog = true
+mon cluster log to syslog = true
+debug mon = 1/5
+debug osd = 1/5
+
+[client]
+log file = /var/log/ceph.log
+"""
+        with open(ceph_conf_path, "w") as f:
+            f.write(ceph_conf)
+
+    def write_ceph_cli_keyring(self) -> None:
+        """Write Ceph CLI keyring file"""
+        ceph_keyring_path = "/etc/ceph/ceph.client.{}.keyring".format(self.app.name)
+        ceph_keyring = f"[client.{self.app.name}]\n\tkey = {self.key}\n"
+        with open(ceph_keyring_path, "w") as f:
+            f.write(ceph_keyring)
+
+    def configure_ceph_cli(self) -> None:
+        """Configure Ceph CLI"""
+        ceph_conf_dir = "/etc/ceph"
+        os.makedirs(ceph_conf_dir, exist_ok=True)
+        self.write_ceph_cli_config()
+        self.write_ceph_cli_keyring()
+
+    def ceph_cli(self, *args: str, timeout: int = 60) -> str:
+        """Run Ceph CLI command"""
+        cmd = ["ceph", "--user", self.app.name] + list(args)
+        return subprocess.check_output(cmd, timeout=timeout).decode("UTF-8")
+
+    def get_ceph_fsid(self) -> str:
+        """Get the Ceph FSID (cluster ID)"""
+        fsid = self.ceph_cli("fsid").strip()
+        return fsid
+
     @property
     def ceph_context(self) -> Dict[str, str]:
         """Return context that can be used to render ceph resource files in templates/ folder."""
-        raw_mon_hosts = self._stored.ceph_data.get("mon_hosts", "").split()
         return {
-            "fsid": self._stored.ceph_data.get("fsid"),
-            "kubernetes_key": self._stored.ceph_data.get("key"),
-            "mon_hosts": json.dumps(raw_mon_hosts),
+            "fsid": self.get_ceph_fsid(),
+            "kubernetes_key": self.key,
+            "mon_hosts": json.dumps(self.mon_hosts),
+            "user": self.app.name,
```

Comment:
so, the template value will change from `admin` to `ceph-csi` or whatever this app is named?  Am i reading that right

Reply:
Yeah, that's right. The `ceph-client` relation gives you an auth key for a user named `<application-name>`. The username isn't in the relation data at all, it's just something you have to know when using the ceph-client relation.